### PR TITLE
add new install bundles api rfc

### PR DESCRIPTION
**Pull request by carneyweb**

### CODE REVIEW

File: text/0006-BundleContext-Install-API.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+* Add **manifest** argument to constructor
+  * If the **manifest** is not empty, use it to initialize the **m_SortedToplevelDirs**, and leave **m_IsContainerOpen** set to false
+    * if **manifest** is empty, no manifest injection is happening, so open up the archive and read the manifest from there, set set **m_IsContainerOpen** to true.
+
+### BundleStorage(Memory)
+
+* Refactor
+  * Remove **InsertArchives()** method and move loop to caller in **BundleRegistry::Install0**
+
```
Comment (jeffdiclemente) [inline text/0006-BundleContext-Install-API.md:120]: a sequence diagram showing how the abovementione classes now interact when `InstallBundles` is called would be helpful to illustrate the proposed changes.

File: text/0006-BundleContext-Install-API.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+```c++
+/** Wrapper around BundleRegistry::Install
+* 
+* @param location The absolute path to the location of the bundle whose manifest 
+*        is to be installed
+* @param manifest the manifest for the bundle. 
+*/
+vector<Bundle> BundleContext::InstallBundles(std::string location
+                                             , std::vector<AnyMap> manifest);
```
Comment (jeffdiclemente) [inline text/0006-BundleContext-Install-API.md:27]: how are statically linked bundles handled? how does someone call this API for a location containing statically linked bundles?

Comment (jeffdiclemente) [inline text/0006-BundleContext-Install-API.md:27]: what happens if the manifest data passed in does not match was is physically on disk at `location`?

File: text/0006-BundleContext-Install-API.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+### Other Changes
+
+* Various code cleanups (getting rid of "using std", exraneous cout output, etc.)
+
+## Test Cases
+
+- Install bundle manifest at a location, and then retrieve bundle and check manifest
+- Install bundle manifest for a test bundle that exists. Then get the service, call a method through the service interface and made sure that it was successfully started.
+- Install bundle manifest for multiple bundles that exist on disk and then start each one and verify that they have been properly instantiated.
+- Ensure Install bundle manifest for bundle already installed is ignored
```
Comment (jeffdiclemente) [inline text/0006-BundleContext-Install-API.md:130]: not sure what is meant by "ignored". A test should guarantee that a bundle installed more than once always returns the originally installed bundle to the caller and does not create multiple entries in the bundle registry for the same bundle.

File: text/0006-BundleContext-Install-API/0006-BundleContext-Install-API.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+Provide a api to initialize the bundle registry with a bundle manifest
+
+## Motivation
+
+Allows users of the system to inject bundle manifests into the registry directly. We will be able to use this mechanism to speed up initial install of very large numbers of bundles.
+
+## Detailed design
+
+![InstallBundlesSequence](./InstallBundlesSequence.svg)
```
Comment (jeffdiclemente) [inline text/0006-BundleContext-Install-API/0006-BundleContext-Install-API.md:17]: link doesn't work. take a look at https://raw.githubusercontent.com/CppMicroServices/rfcs/master/text/0003-declarative-services.md for an example of how .svg files are added to a document.

**Comment by carneyweb** [comment]
Thanks. Hadn't had a chance to try and look at it online yet. Will fix and re-upload

- Mike.

On July 6, 2020 at 11:56 AM, Jeff DiClemente <notifications@github.com> wrote:


@jeffdiclemente commented on this pull request.
In text/0006-BundleContext-Install-API/0006-BundleContext-Install-API.md:
> +- RFC PR: (in a subsequent commit to the PR, fill me with the PR's URL)
+- CppMicroServices Issue: (fill me with the URL of the CppMicroServices issue that necessisated this RFC, otherwise leave blank)
+
+# Add BundleContext::InstallBundles API
+
+## Summary
+
+Provide a api to initialize the bundle registry with a bundle manifest
+
+## Motivation
+
+Allows users of the system to inject bundle manifests into the registry directly. We will be able to use this mechanism to speed up initial install of very large numbers of bundles.
+
+## Detailed design
+
+InstallBundlesSequence
link doesn't work. take a look at https://raw.githubusercontent.com/CppMicroServices/rfcs/master/text/0003-declarative-services.md for an example of how .svg files are added to a document.

File: text/0006-BundleContext-Install-API.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+### BundleStorage(Memory)
+
+* Refactor
+  * Remove **InsertArchives()** method and move loop to caller in **BundleRegistry::Install0**
+
+### Other Changes
+
+* Various code cleanups (getting rid of "using std", exraneous cout output, etc.)
+
+## Test Cases
```
[the remark this answers is not in the document]
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:138]: Added back in.

File: text/0006-BundleContext-Install-API.md (hunk 1 above)
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:133]: Done.

File: text/0006-BundleContext-Install-API.md (hunk 2 above)
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:39]: The behavior is undefined (the manifests are never compared... the system assumes that they match, and if they don't who knows what's going to happen). I added a remark in the Doxygen comment explaining this.

File: text/0006-BundleContext-Install-API.md (hunk 3 above)
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:143]: Updated description of this test case to match your wording.

File: text/0006-BundleContext-Install-API.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+# Add BundleContext::InstallBundles API
+
+## Summary
+
+Provide a api to initialize the bundle registry with a bundle manifest
+
+## Motivation
+
+Allows users of the system to inject bundle manifests into the registry directly. We will be able to use this mechanism to speed up initial install of very large numbers of bundles.
+
```
[the remark this answers is not in the document]
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:14]: Added first draft of a motivation use case. See if the basic idea works for you. If so, I can iterate on the presentation, and if not I'll try and think of something else suitably generic but applicable enough :-)

File: text/0006-BundleContext-Install-API.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
+- Start Date: 2020-06-09
+- RFC PR: (in a subsequent commit to the PR, fill me with the PR's URL)
+- CppMicroServices Issue: (fill me with the URL of the CppMicroServices issue that necessisated this RFC, otherwise leave blank)
+
+# Add BundleContext::InstallBundles API
```
[the remark this answers is not in the document]
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md]: Updated to "Add BundleManifest Injection Capabilities"

File: text/0006-BundleContext-Install-API.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+### BundleStorage(Memory)
+
+* Refactor
+  * Remove **InsertArchives()** method and move loop to caller in **BundleRegistry::Install0**
+
+### Other Changes
+
+* Various code cleanups (getting rid of "using std", exraneous cout output, etc.)
+
```
[the remark this answers is not in the document]
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:166]: Added the section back as a reminder that I need to fill it in.

[the remark this answers is not in the document]
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:166]: Added.

File: text/0006-BundleContext-Install-API.md (hunk 2 above)
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:68]: There's a description in the "How do we teach this" section of the doc, and a "TODO" comment in the actual header.

Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:68]: I added a note to this effect in the doc as well.

File: text/0006-BundleContext-Install-API.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+- RFC PR: (in a subsequent commit to the PR, fill me with the PR's URL)
+- CppMicroServices Issue: (fill me with the URL of the CppMicroServices issue that necessisated this RFC, otherwise leave blank)
+
+# Add BundleManifest Injection Capabilities 
+
+## Summary
+
+Provide a api to initialize the bundle registry with a bundle manifest during the BundleContext::InstallBundles operation by providing an optional argument containing a manifest.
+
+## Motivation
```
Comment (jeffdiclemente) [inline text/0006-BundleContext-Install-API.md:11]: somewhere it should be noted that this design attempts to follow the OSGi spec by implementing an equivalence of https://docs.osgi.org/specification/osgi.core/7.0.0/framework.api.html#org.osgi.framework.BundleContext.installBundle-String-InputStream-

Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:11]: Added.

File: text/0006-BundleContext-Install-API.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+This proposal discusses an approach that could solve this problem with an API for injecting manifests into the BundleRegistry directly, without opening and reading the manifests from the bundles.
+
+## Detailed design
+
+### Class Diagram
+
+This is a partial class diagram showing the relationship between the pieces of the system used during the BundleContext::InstallBundles operation. Only those classes and fields that participate in that operation are shown.
+
+![ClassDiagram](0006-BundleContext-Install-API/ClassDiagram.svg)
```
[the remark this answers is not in the document]
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:27]: Please try again. I can see both images in the document at this link:

https://github.com/CppMicroServices/rfcs/blob/newInstallAPI/text/0006-BundleContext-Install-API.md

File: text/0006-BundleContext-Install-API.md [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+
+### Sequence
+
+![InstallBundlesSequence](0006-BundleContext-Install-API/InstallBundlesSequence.svg)
```
[the remark this answers is not in the document]
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:31]: See above. Check this page: https://github.com/CppMicroServices/rfcs/blob/newInstallAPI/text/0006-BundleContext-Install-API.md

File: text/0006-BundleContext-Install-API.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+# Add BundleManifest Injection Capabilities 
+
+## Summary
+
+Provide a api to initialize the bundle registry with a bundle manifest during the BundleContext::InstallBundles operation by providing an optional argument containing a manifest.
+
+## Motivation
+
+Allows users of the system to install bundle manifests into the registry directly. We will be able to use this mechanism to speed up initial install of very large numbers of bundles.
+
```
[the remark this answers is not in the document]
Comment (carneyweb) [reply text/0006-BundleContext-Install-API.md:14]: Changed the first sentence to "Allows users of the system to install bundle without opening the archive containing the bundle."

**Pull request merged**